<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
</head>
<body>

<script type="module">
    import KPromise from './src/KPromise.js';

    //// test 测试KPromise的then、catch方法 start

    let p = new KPromise((resolve, reject) => {
        let num = Math.random();
        setTimeout(() => {
            num > 0.5 ? resolve('hello') : reject('报错啦');
        }, 2000);
    });

    p.then(res => {
        console.log('1、'+res);
        return `then1:${res}`;
    }).then(res => {
        console.log('2、'+res);
        return `then2:${res}`;
    }, reason => {
        console.log(`3、onRejected----->${reason}`);
    });
    setTimeout(() => {
        p.then(res => console.log('4、--->', res));
    }, 3000);

    //// test 测试KPromise的then、catch方法 end

    //// test 测试静态方法resolve、reject、all、race    start
    KPromise.resolve('你好').then(res=>{
        console.log(`%c5、静态方法：KPromise.resolve调用输出：${res}`,'color:green;');
    });
    KPromise.reject('这是个错误').catch(res=>{
        console.log(`%c6、静态方法：KPromise.reject调用输出：${res}`,'color: red;');
    });

    let p1 = new KPromise((resolve, reject)=>{
        setTimeout(()=>{
            resolve('p1完成');
        },1000);
    });

    let p2 = new KPromise((resolve, reject)=>{
        let num = Math.random();
        setTimeout(()=>{
            num<0.5?reject('p2报错'):resolve('p2成功');
        },500);
    });
    let p3 = new KPromise((resolve, reject)=>{
        setTimeout(()=>{
            resolve('p3完成');
        },500);
    });

    KPromise.all([p1,p2,p3]).then(res=>console.log('7、KPromise.all:',res)).catch(reason=>console.log(`7、KPromise.all报错:${reason}`));
    KPromise.race([p1,p2,p3]).then(res=>console.log('8、KPromise.race:',res)).catch(reason=>console.log(`8、KPromise.race报错:${reason}`));
    //// test 测试静态方法resolve、reject、all、race    end

    //// test async-await调用方式测试 start
    (async ()=>{
        try{
            let res = await new KPromise((resolve, reject) => {
                let num = Math.random();
                setTimeout(()=>{

                    num>0.5?resolve('hello'):reject('报错啦');
                },2000);

            }).then(res=>`9、then1:${res}`);
        }catch (e) {
            console.log('9、async 报错了',e);
        }
    })();
    //// test async-await调用方式测试 end
</script>

</body>
</html>